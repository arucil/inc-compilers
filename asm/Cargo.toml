[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
