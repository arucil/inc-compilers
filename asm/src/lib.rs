#![feature(never_type)]

use num_derive::{FromPrimitive, ToPrimitive};
use std::fmt::{self, Debug, Formatter, Write};

#[derive(Debug, Clone)]
pub struct Program<INFO = (), VAR = !> {
  pub info: INFO,
  pub blocks: Vec<(String, Block<VAR>)>,
}

#[derive(Clone)]
pub struct Block<VAR = !> {
  pub code: Vec<Instr<VAR>>,
}

#[derive(Clone)]
#[non_exhaustive]
pub enum Instr<VAR = !> {
  Add(Arg<VAR>, Arg<VAR>),
  Sub(Arg<VAR>, Arg<VAR>),
  Mov(Arg<VAR>, Arg<VAR>),
  Neg(Arg<VAR>),
  Call(String, usize),
  Ret,
  Push(Arg<VAR>),
  Pop(Arg<VAR>),
  Jmp(String),
  Syscall,
}

#[derive(Clone, PartialEq, Eq)]
pub enum Arg<VAR = !> {
  Imm(i64),
  Reg(Reg),
  Deref(Reg, i32),
  Var(VAR),
}

#[derive(
  Clone, Copy, FromPrimitive, ToPrimitive, PartialEq, Eq, PartialOrd, Ord, Hash,
)]
pub enum Reg {
  Rsp,
  Rbp,
  Rax,
  Rbx,
  Rcx,
  Rdx,
  Rsi,
  Rdi,
  R8,
  R9,
  R10,
  R11,
  R12,
  R13,
  R14,
  R15,
}

impl<INFO: Debug, VAR: Debug> Program<INFO, VAR> {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = format!("{:?}", self.info);
    for (label, block) in &self.blocks {
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }

  pub fn to_nasm(&self) -> String {
    let mut buf = format!(
      "extern read_int, print_int, print_newline\n\
      section .text\n"
    );
    for (label, block) in &self.blocks {
      buf += "\n";
      if label == "_start" {
        buf += "    global _start\n";
      }
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }
}

impl<VAR: Debug> Debug for Block<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    for instr in &self.code {
      write!(f, "    ")?;
      instr.fmt(f)?;
      writeln!(f)?;
    }
    Ok(())
  }
}

impl<VAR: Debug> Debug for Instr<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Add(src, dest) => write!(f, "add {:?}, {:?}", dest, src),
      Self::Mov(src, dest) => write!(f, "mov {:?}, {:?}", dest, src),
      Self::Call(label, _) => write!(f, "call {}", label),
      Self::Jmp(label) => write!(f, "jmp {}", label),
      Self::Neg(dest) => write!(f, "neg {:?}", dest),
      Self::Pop(dest) => write!(f, "pop {:?}", dest),
      Self::Push(src) => write!(f, "push {:?}", src),
      Self::Ret => write!(f, "ret"),
      Self::Sub(src, dest) => write!(f, "sub {:?}, {:?}", dest, src),
      Self::Syscall => write!(f, "syscall"),
    }
  }
}

impl<VAR: Debug> Debug for Arg<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Imm(n) => write!(f, "{}", n),
      Self::Var(var) => var.fmt(f),
      Self::Deref(r, i) => {
        if *i > 0 {
          write!(f, "qword [{:?} + {}]", r, i)
        } else if *i == 0 {
          write!(f, "qword [{:?}]", r)
        } else {
          write!(f, "qword [{:?} - {}]", r, -i)
        }
      }
      Self::Reg(r) => r.fmt(f),
    }
  }
}

impl Reg {
  pub fn is_callee_saved(&self) -> bool {
    match self {
      Reg::Rsp
      | Reg::Rbp
      | Reg::Rbx
      | Reg::R12
      | Reg::R13
      | Reg::R14
      | Reg::R15 => true,
      _ => false,
    }
  }
}

impl Debug for Reg {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    use Reg::*;
    let op = match self {
      Rsp => "rsp",
      Rbp => "rbp",
      Rax => "rax",
      Rbx => "rbx",
      Rcx => "rcx",
      Rdx => "rdx",
      Rsi => "rsi",
      Rdi => "rdi",
      R8 => "r8",
      R9 => "r9",
      R10 => "r10",
      R11 => "r11",
      R12 => "r12",
      R13 => "r13",
      R14 => "r14",
      R15 => "r15",
    };
    f.write_str(op)
  }
}
