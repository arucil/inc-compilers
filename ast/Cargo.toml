[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
