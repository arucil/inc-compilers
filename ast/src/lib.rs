#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use pretty::*;
use std::fmt::{self, Debug, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub body: Vec<Exp<VAR, TYPE>>,
  pub types: Arena<CompType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Type {
  Int,
  Bool,
  Str,
  Comp(TypeId),
  Void,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TypeId(Id<CompType>);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompType {
  Vector(Vec<Type>),
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl<VAR: Debug, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(30, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(self.body.iter().map(|exp| exp.to_doc()), Doc::line())
  }
}

impl<VAR: Debug, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) | ExpKind::Get(var) => {
        RcDoc::text(format!("{:?}", var))
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{:?}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append(RcDoc::text("]")),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc().append(Doc::line()).append(alt.to_doc()))
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(RcDoc::text(format!("{:?}", var.1)))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin").append(Doc::line()).append(
            RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            )
            .nest(1)
            .group(),
          ),
        )
        .append(RcDoc::text(")")),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::NewLine => RcDoc::text("(newline)"),
    }
  }
}
