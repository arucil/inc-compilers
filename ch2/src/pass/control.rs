use ast::{Program, Exp, IdxVar};
use std::fmt::{self, Write};
use indexmap::IndexSet;
use support::{WritePretty, Range};

pub struct CProgram {
  pub info: CInfo,
  pub body: Vec<(String, CTail)>,
}

pub struct CInfo {
  pub locals: IndexSet<IdxVar>,
}

pub enum CTail {
  Return(CExp),
  Seq(CStmt, Box<CTail>)
}

pub enum CStmt {
  Assign {
    var: IdxVar,
    exp: CExp,
  }
}

pub enum CExp {
  Atom(CAtom),
  Prim(CPrim),
}

pub enum CPrim {
  Read,
  Neg(CAtom),
  Add(CAtom, CAtom),
}

pub enum CAtom {
  Int(i64),
  Var(IdxVar)
}

impl CProgram {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = String::new();
    self.info.write(&mut buf).unwrap();
    for (label, tail) in &self.body {
      writeln!(&mut buf, "{}:", label).unwrap();
      tail.write(&mut buf).unwrap();
    }
    buf
  }
}

impl WritePretty for CInfo {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

impl WritePretty for CTail {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Return(exp) => {
        write!(f, "    return ")?;
        exp.write(f)
      }
      Self::Seq(stmt, tail) => {
        write!(f, "    ")?;
        stmt.write(f)?;
        write!(f, "\n")?;
        tail.write(f)
      }
    }
  }
}

impl WritePretty for CStmt {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Assign { var, exp } => {
        write!(f, "{:?} = ", var)?;
        exp.write(f)
      }
    }
  }
}

impl WritePretty for CExp {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Atom(atom) => atom.write(f),
      Self::Prim(prim) => prim.write(f),
    }
  }
}

impl WritePretty for CPrim {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Add(arg1, arg2) => {
        write!(f, "(+ ")?;
        arg1.write(f)?;
        write!(f, " ")?;
        arg2.write(f)?;
        write!(f, ")")
      }
      Self::Neg(arg) => {
        write!(f, "(- ")?;
        arg.write(f)?;
        write!(f, ")")
      }
      Self::Read => {
        write!(f, "(read)")
      }
    }
  }
}

impl WritePretty for CAtom {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Int(n) => write!(f, "{}", n),
      Self::Var(n) => write!(f, "{:?}", n),
    }
  }
}

pub fn explicate_control(mut prog: Program<IdxVar>) -> CProgram {
  let locals = collect_locals(&prog);
  CProgram {
    info: CInfo {
      locals,
    },
    body: vec![
      ("start".to_owned(), explicate_tail(prog.body.pop().unwrap().1))
    ]
  }
}

fn collect_locals(
  prog: &Program<IdxVar>,
) -> IndexSet<IdxVar> {
  let mut locals = IndexSet::new();
  for (_, exp) in &prog.body {
    collect_exp_locals(exp, &mut locals);
  }
  locals
}

fn collect_exp_locals(
  exp: &Exp<IdxVar>,
  locals: &mut IndexSet<IdxVar>,
) {
  match exp {
    Exp::Int(_) => {}
    Exp::Var(var) => {
      locals.insert(var.clone());
    }
    Exp::Str(_) => {}
    Exp::Let { var, init, body } => {
      locals.insert(var.1.clone());
      collect_exp_locals(&init.1, locals);
      collect_exp_locals(&body.1, locals);
    }
    Exp::Prim { op: _, args } => {
      for (_, arg) in args {
        collect_exp_locals(arg, locals);
      }
    }
    _ => unimplemented!()
  }
}

fn explicate_tail(exp: Exp<IdxVar>) -> CTail {
  match exp {
    exp@(Exp::Int(_) | Exp::Var(_)) => {
      CTail::Return(CExp::Atom(atom(exp)))
    }
    ast::Exp::Prim { op: (_, op), args } => {
      CTail::Return(CExp::Prim(prim(op, args)))
    }
    ast::Exp::Let { var, init, body } => {
      explicate_assign(var.1, init.1, explicate_tail(body.1))
    }
    exp => unimplemented!("unsupported form {:?}", exp)
  }
}

fn explicate_assign(
  var: IdxVar,
  init: ast::Exp<IdxVar>,
  cont: CTail,
) -> CTail {
  match init {
    exp@(Exp::Int(_) | Exp::Var(_)) => {
      let assign = CStmt::Assign {
        var,
        exp: CExp::Atom(atom(exp)),
      };
      CTail::Seq(assign, box cont)
    }
    Exp::Prim { op: (_, op), args } => {
      let prim = prim(op, args);
      let assign = CStmt::Assign {
        var,
        exp: CExp::Prim(prim),
      };
      CTail::Seq(assign, box cont)
    }
    Exp::Let { var: (_, var1), init: box (_, init1), body } => {
      explicate_assign(var1, init1, explicate_assign(var, body.1, cont))
    }
    exp => unimplemented!("unsupported form {:?}", exp)
  }
}

fn atom(exp: Exp<IdxVar>) -> CAtom {
  match exp {
    Exp::Int(n) => CAtom::Int(n),
    Exp::Var(var) => CAtom::Var(var),
    exp => unreachable!("{:?}", exp),
  }
}

fn prim(op: &str, mut args: Vec<(Range, Exp<IdxVar>)>) -> CPrim {
  match op {
    "read" => CPrim::Read,
    "-" => CPrim::Neg(atom(args.pop().unwrap().1)),
    "+" => CPrim::Add(
      atom(args[0].1.clone()),
      atom(args.pop().unwrap().1),
    ),
    _ => unreachable!("{}", op)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn let_in_init() {
    let prog = parse(r#"(let ([y (let ([x 20]) (+ x (let ([x 22]) x)))]) y)"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_prims() {
    let prog = parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}