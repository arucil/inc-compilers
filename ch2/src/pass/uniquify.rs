use ast::{Exp, IdxVar, Program};
use std::collections::HashMap;
use support::{CompileError, Range};

/// Make variable names unique.
pub fn uniquify(prog: Program) -> Result<Program<IdxVar>, CompileError> {
  let mut counter = 0;
  let mut env = HashMap::new();
  Ok(Program {
    body: prog
      .body
      .into_iter()
      .map(|exp| {
        Ok((
          exp.0,
          uniquify_exp(uniquify_exp, exp, &mut env, &mut counter)?,
        ))
      })
      .collect::<Result<_, _>>()?,
  })
}

pub type uniquify_exp_fn = fn(
  uniquify_exp_fn,
  (Range, Exp),
  &mut HashMap<String, usize>,
  &mut usize,
) -> Result<Exp<IdxVar>, CompileError>;

pub fn uniquify_exp(
  recur: uniquify_exp_fn,
  (range, exp): (Range, Exp),
  env: &mut HashMap<String, usize>,
  counter: &mut usize,
) -> Result<Exp<IdxVar>, CompileError> {
  match exp {
    Exp::Int(n) => Ok(Exp::Int(n)),
    Exp::Var(var) => env.get(&var).map_or_else(
      || {
        Err(CompileError {
          range,
          message: format!("variable {} not found", var),
        })
      },
      |&index| {
        Ok(Exp::Var(IdxVar {
          name: var.clone(),
          index,
        }))
      },
    ),
    Exp::Prim { op, args } => Ok(Exp::Prim {
      op,
      args: args
        .into_iter()
        .map(|exp| Ok((exp.0, recur(recur, exp, env, counter)?)))
        .collect::<Result<_, _>>()?,
    }),
    Exp::Let {
      var: var @ (var_range, _),
      init: box init @ (init_range, _),
      body: box body @ (body_range, _),
    } => {
      let init = recur(recur, init, env, counter)?;
      let index = *counter;
      *counter += 1;
      let old_value = env.insert(var.1.clone(), index);
      let body = recur(recur, body, env, counter)?;
      if let Some(v) = old_value {
        env.insert(var.1.clone(), v);
      }
      Ok(Exp::Let {
        var: (var_range, IdxVar { name: var.1, index }),
        init: box (init_range, init),
        body: box (body_range, body),
      })
    }
    exp => unimplemented!("unsupported form {:?}", exp),
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn let_form() {
    let prog = parse(r#"(let ([x (- (read))]) (+ 2 x))"#).unwrap();
    let result = uniquify(prog).unwrap();
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_form_in_init() {
    let prog = parse(r#"(let ([x (let ([y 3]) y)]) (+ 2 x))"#).unwrap();
    let result = uniquify(prog).unwrap();
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_form_in_init_shadows() {
    let prog = parse(r#"(let ([x (let ([x 3]) x)]) (+ 2 x))"#).unwrap();
    let result = uniquify(prog).unwrap();
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_form_in_body_shadows() {
    let prog = parse(
      r#"(let ([x (let ([x 3]) x)]) (+ (let ([x 3] [y (read)]) (+ x y)) x))"#,
    )
    .unwrap();
    let result = uniquify(prog).unwrap();
    assert_snapshot!(result.to_string_pretty());
  }
}
