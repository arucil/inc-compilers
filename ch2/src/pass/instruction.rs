use super::control::{CExp, CProgram, CStmt, CTail, CAtom, CPrim};
use super::uniquify::IdxVar;
use asm::{Program, Block, Instr, Arg, Reg};
use std::iter;

pub fn select_instruction(prog: CProgram) -> Program<IdxVar> {
  Program {
    blocks: prog.body.into_iter()
      .map(|(label, tail)| (label, tail_block(tail)))
      .collect()
  }
}

fn tail_block(tail: CTail) -> Block<IdxVar> {
  let mut code = vec![];
  tail_instructions(tail, &mut code);
  Block {
    code,
  }
}

fn tail_instructions(
  mut tail: CTail,
  code: &mut Vec<Instr<IdxVar>>,
) {
  loop {
    match tail {
      CTail::Return(exp) => {
        exp_instructions(Arg::Reg(Reg::Rax), exp, code);
        code.push(Instr::Jmp("conclusion".to_owned()));
        return;
      }
      CTail::Seq(stmt, new_tail) => {
        stmt_instructions(stmt, code);
        tail = *new_tail;
      }
    }
  }
}

fn stmt_instructions(
  stmt: CStmt,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match stmt {
    CStmt::Assign { var, exp } => exp_instructions(Arg::Var(var), exp, code),
  }
}

fn exp_instructions(
  target: Arg<IdxVar>,
  exp: CExp,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match exp {
    CExp::Atom(atom) => {
      atom_instructions(target, atom, code);
    }
    CExp::Prim(CPrim::Read) => {
      code.push(Instr::Callq("read_int".to_owned(), 0));
      code.push(Instr::Movq(Arg::Reg(Reg::Rax), target));
    }
    CExp::Prim(CPrim::Neg(atom)) => {
      atom_instructions(target.clone(), atom, code);
      code.push(Instr::Negq(target));
    }
    CExp::Prim(CPrim::Add(atom1, atom2)) => {
      atom_instructions(target.clone(), atom1, code);
      let arg = match atom2 {
        CAtom::Int(n) => Arg::Imm(n),
        CAtom::Var(var) => Arg::Var(var),
      };
      code.push(Instr::Addq(arg, target));
    }
  }
}

fn atom_instructions(
  target: Arg<IdxVar>,
  atom: CAtom,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match atom {
    CAtom::Int(n) => {
      code.push(Instr::Movq(Arg::Imm(n), target));
    }
    CAtom::Var(var) => {
      code.push(Instr::Movq(Arg::Var(var), target));
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog = parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let prog = super::super::control::explicate_control(prog);
    let result = select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}