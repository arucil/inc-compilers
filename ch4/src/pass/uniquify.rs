use ast::{IdxVar, Program};
use ch2::pass::uniquify::uniquify_exp;
use std::collections::HashMap;

pub fn uniquify<TYPE>(prog: Program<String, TYPE>) -> Program<IdxVar, TYPE> {
  let mut counter = 0;
  let mut env = HashMap::new();
  Program {
    body: prog
      .body
      .into_iter()
      .map(|exp| uniquify_exp(exp, &mut env, &mut counter).unwrap())
      .collect(),
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn let_form() {
    let prog = parse(r#"(let ([x (- (read))]) (+ 2 x))"#).unwrap();
    let result = uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_form_in_init() {
    let prog = parse(r#"(let ([x (let ([y 3]) y)]) (+ 2 x))"#).unwrap();
    let result = uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_form_in_init_shadows() {
    let prog = parse(r#"(let ([x (let ([x 3]) x)]) (+ 2 x))"#).unwrap();
    let result = uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_form_in_body_shadows() {
    let prog = parse(
      r#"(let ([x (let ([x 3]) x)]) (+ (let ([x 3] [y (read)]) (+ x y)) x))"#,
    )
    .unwrap();
    let result = uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_form() {
    let prog =
      parse(r#"(let ([x (read)]) (if (eq? x 3) (> 2 1) (not x)))"#).unwrap();
    let result = uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
