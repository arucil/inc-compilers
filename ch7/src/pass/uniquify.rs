#[cfg(test)]
mod tests {
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn params() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  (+ x y))
(define (bar [x Int] [y Int]) : Int
  (let ([x (+ x 1)] [f bar])
    (set! f foo)
    (* x y)))
(let ([x (if #t foo bar)])
  (print (x 1 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let result = uniquify::uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
