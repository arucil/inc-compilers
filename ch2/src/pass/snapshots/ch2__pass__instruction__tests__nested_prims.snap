---
source: ch2/src/pass/instruction.rs
expression: result.to_string_pretty()

---
start:
    callq read_int
    movq rax, x.0
    movq 2, y.1
    addq 3, y.1
    callq read_int
    movq rax, tmp.0
    movq tmp.0, tmp.1
    negq tmp.1, 
    movq 2, tmp.2
    negq tmp.2, 
    movq y.1, tmp.3
    addq tmp.2, tmp.3
    movq tmp.1, rax
    addq tmp.3, rax
    jmp conclusion

